kernel/net/src/device.rs

```diff
@@ -0,0 +1,113 @@
+use alloc::vec;
+use core::any::Any;
+use log::error;
+use nic_buffers::ReceivedFrame;
+use smoltcp::phy;
+
+pub use phy::DeviceCapabilities;
+
+use crate::Error;
+
+/// Standard maximum transition unit for ethernet cards.
+const STANDARD_MTU: usize = 1500;
+
+/// A network device.
+///
+/// Devices implementing this trait can then be registered using
+/// [`register_device`].
+///
+/// [`register_device`]: crate::register_device
+pub trait Device: Send + Sync + Any {
+    fn send(&mut self, buf: &[u8]) -> core::result::Result<(), crate::Error>;
+
+    fn receive(&mut self) -> Option<ReceivedFrame>;
+
+    /// Returns the MAC address.
+    fn mac_address(&self) -> [u8; 6];
+
+    /// Returns the device capabilities.
+    fn capabilities(&self) -> DeviceCapabilities {
+        let mut caps = DeviceCapabilities::default();
+        caps.max_transmission_unit = STANDARD_MTU;
+        caps
+    }
+}
+
+/// Wrapper around a network device.
+///
+/// We use this because we can't directly implement [`phy::Device`] for `T` due
+/// to the following error:
+/// ```
+/// error[E0210]: type parameter `T` must be used as the type parameter for some local type (e.g., `MyStruct<T>`)
+/// ```
+pub(crate) struct DeviceWrapper<'a> {
+    pub(crate) inner: &'a mut dyn Device,
+}
+
+impl<'a, 'b> phy::Device<'a> for DeviceWrapper<'b> {
+    type RxToken = RxToken;
+
+    type TxToken = TxToken<'a>;
+
+    fn receive(&'a mut self) -> Option<(Self::RxToken, Self::TxToken)> {
+        let frame = self.inner.receive()?;
+        Some((RxToken { inner: frame }, TxToken { device: self.inner }))
+    }
+
+    fn transmit(&'a mut self) -> Option<Self::TxToken> {
+        Some(TxToken { device: self.inner })
+    }
+
+    fn capabilities(&self) -> DeviceCapabilities {
+        self.inner.capabilities()
+    }
+}
+
+/// The receive token.
+pub(crate) struct RxToken {
+    inner: ReceivedFrame,
+}
+
+impl phy::RxToken for RxToken {
+    fn consume<R, F>(mut self, _timestamp: smoltcp::time::Instant, f: F) -> smoltcp::Result<R>
+    where
+        F: FnOnce(&mut [u8]) -> smoltcp::Result<R>,
+    {
+        if self.inner.0.len() > 1 {
+            error!(
+                "DeviceWrapper::receive(): received frame with {} buffers",
+                self.inner.0.len()
+            );
+        }
+        // TODO: Which error variant?
```

Comment:
`Exhausted` i believe.

It could also be `Finished`, though I don't think we have sufficient information here to determine whether the remote end was closed